### ISSUE TITLE
Magic Replace prompt text in the Layer panel for a Magic layer changes unexpectedly

### ISSUE THREAD

**Issue by lardofdorkness** [photopea/photopea#6888]
Your Operating System: macOS Monterey 12.7.5
Your Web Browser: Google Chrome
The problem happens in the Incognito Mode, too: unknown

**This must be done to see the problem**
1. Go to www.photopea.com
2. Create a new or open an existing project.
3. Select a region of pixels, then select the Magic Replace tool.
4. Enter something for the prompt, then click Replace.
5. A new Magic Layer is created and selected, and the Layer panel opens to reveal the Magic Replace controls for the layer, including the text of the Prompt that was just entered.
6. Now change the Prompt text in the Layer panel and click Generate.
7. When the action begins to Process, the Prompt text reverts to what had previously been entered.
8. When the Magic Replace action completes, the Prompt text returns to the latest prompt (i.e., the one that was just processed)

It is strange to see the prompt text in the Layer panel change when I click Generate. It does process the prompt that was last entered, but it also makes me question what I just entered.

I've also noticed that if I'm on a Magic layer that is visible, and I change the prompt text, then turn off visibility for the layer, the prompt text reverts to the prompt for the generation that is selected. This doesn't make sense, although I see why it does make sense when making the layer visible when it previously wasn't. If I'm hiding the layer (so that its content is not recognized by the Magic Replace action), the prompt text shouldn't change as a result.

**Comment by photopea** [comment]
The problem with changing the prompt "temporarily" is fixed now.

The properties window shows you multiple versions of a content of the magic layer. If you select a version, the "Prompt" will contain the prompt of that version. Typing a different prompt will be ignored. Unless you click "Generate" to use your text as a Prompt for a new version.

**Comment by lardofdorkness** [comment]
I appreciate you fixing the first problem. Yes, the prompt may have only changed temporarily, but it was weird for it to do that. :)

As for the second problem I described, try this:

1. Go to www.photopea.com
2. Create a new or open an existing project.
3. Select a region of pixels, then select the Magic Replace tool.
4. Enter something for the prompt, then click Replace.
5. A new Magic Layer is created and selected as the active layer, and the Properties panel opens to reveal the Magic Replace controls for the layer, including the text of the Prompt that was just entered. The layer content is visible by default, and there will be just one version of the prompt result showing below the Magic Replace inputs.
6. Now change the Prompt text in the Properties panel.
7. Next, turn off the Magic layer's visibility by clicking the eye icon for that layer. The prompt text reverts to the prompt text for the selected (sole) version.

This seems problematic, because I haven't changed the selected version, I've only changed the visibility for the selected Magic layer. I would not expect that action to wipe out/replace the new text I entered. If I clicked Generate without changing the visibility, it would process the prompt without changing it. I do understand that switching between versions updates the prompt to the one that generated selected version, but that isn't what I'm describing.

It does seem correct that if the visibility is turned off, and I alter the text, then turn visibility back on, the prompt text should revert to the prompt for the selected version (as it currently does). It's only when I'm turning off visibility that I think it should leave the prompt text alone.

**Comment by photopea** [comment]
You can not change the prompt of a selected version. This is an expected behaviour. Unless you click "Generate", it will be replaced back with the prompt of a selected version at some point.

**Comment by lardofdorkness** [comment]
You're missing my point, and I feel like we're talking past one another. I'm updating the prompt within the Properties panel in order to generate a new version. There will always be a version selected, but in order to generate a new version, I type new text into the Prompt text box and click Generate, and it generates a new version with my prompt. That's not me trying to change the prompt of a selected version, it's just me updating the prompt to create a new one. Good so far?

Now, please try what I suggested in the previous comment. With a Magic layer selected and visible, and a generated version selected, modify the prompt text. BEFORE clicking Generate, turn off visibility for the layer. This action causes the prompt text to revert to that of the selected version, and those actions seem unrelated -- toggling a layer's visibility to OFF shouldn't change what I entered in the prompt. I understand that the prompt can't change for versions that are already generated, but that's not what I'm trying to do and not what's going on. What's happening is it is replacing the prompt that I modified with the prompt from a previous version because I took the unrelated action of turning off visibility. 

I don't know how to explain it any more clearly. Please make it so that it doesn't overwrite new text that I've typed in the prompt input box when I turn off the Magic layer's visibility.

**Comment by photopea** [comment]
I understand it. But how long do you want your new prompt to stay there? Do you want to be able to close and reopen Photopea, and still see your prompt? 

Right now, if you type a new prompt, and click anywhere outside thext area (layer visibility in your case), the "prompt" text area will be updated with the original prompt of the selected version, and you will lose your new prompt.

If you want to compose your prompt while using Photopea, compose it outside Photopea and then, copy-paste it into Photopea, and click "Generate".

**Comment by photopea** [comment]
I could create two text boxes instead of one:

"Prompt of a current version:" - not editable, updated by Photopea
"Prompt for a new version:" - editable, used for "Generate"

But everyone seems to be fine with the way it works now

**Comment by lardofdorkness** [comment]
So, you'd rather complicate the interface by creating a whole separate text box than simply decouple the resetting of the prompt text from the visibility toggle?

No, I don't want to be able to close and reopen Photopea and still see the prompt. You're hyperbolizing my point. The prompt text box in the Properties panel appears to serves two purposes: It can reflect the prompt for a previously-generated version whenever a different version is selected, but it's also the input for generating new versions, correct? So if enter new text there, but then realize that I want the layer to be hidden before I click Generate, it's annoying to have my new text wiped out when I hide the layer. That action shouldn't affect what I've just entered, even if every other action does. (Note: If the layer was already hidden, and I made it visible again after entering new prompt text, I _would_ expect the prompt text I'd entered to be overwritten with the prompt for whichever version was already selected in the Properties pane -- that _does_ make sense)

> Right now, if you type a new prompt, and click anywhere outside thext area (layer visibility in your case), the "prompt" text area will be updated with the original prompt of the selected version, and you will lose your new prompt.

This seems to be what was behind the other issue I reported at the top of this thread: When I would click Generate after updating the prompt, the text would be overwritten with the prompt for the selected version while the request processed, then revert to the prompt that was last entered and processed. Since clicking anywhere outside the text area updates the prompt text, submitting the form had the same effect, which was an odd thing to have happen, even if it was "temporary." In that particular case, you were able to decouple the updating of the prompt text from the form submit action. I don't understand why you cannot do the same for when visibility is turned off for the selected Magic Layer.

This is a minor issue, and certainly not worth the amount of time I've put into explaining it and you into defending it. To me, even small software quirks detract from an application's usability and my confidence in it, and therefore merit attention. Maybe "everyone" seems fine with this because it's less taxing to ignore or work around such issues than report them. If the behavior is what it is and I can't convince you to make the simple change I requested, I can instead be careful to hide the layer first, before updating the prompt text. You must know that it wouldn't make sense to have to use a separate application to compose prompt text, or for there to be another text box in the Magic Replace form.
